### PR TITLE
Update to fix late validation plots  from PR 15562

### PR DESCRIPTION
**Pull request by violatingcp**
Following concerns in    #15562 we have fix based on the validation plots. This results from the fact that when CandViewMerger is used, the referening of the packedcandidates to the orignal leptons is somehow broken. PuppiPhoton relies on these references to remove the give pfcandidates assigned to a good photon weight 1. If this is broken Puppi photon will add back the missing candidates even if they are there. The solution is to either run puppi no lepton or puppiphoton, but not both. The problem with this is that puppi vs puppi no lepton will reduce the MET response by a 2-3% whereas puppiphoton vs puppi will make puppiMET robust in photon+jet events. We prfer to have puppi photon. On another note, after miniAOD is run, the puppinolepton weights attached to the packed candidates give weight 1 for the leptons, so puppiphoton can be run on the miniAOD using the precomputed puppinolepton weights allowing for the best puppiMET peformance.

### CODE REVIEW

File: PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py [hunk 1]
```diff
@@ -21,8 +21,8 @@ def makePuppies( process ):
     process.puppiMerged = cms.EDProducer("CandViewMerger",src = cms.VInputTag( 'puppiNoLep','pfLeptonsPUPPET'))
     process.load('CommonTools.PileupAlgos.PhotonPuppi_cff')
     process.puppiForMET = process.puppiPhoton.clone()
-    process.puppiForMET.puppiCandName    = 'puppiMerged'
-
+    #process.puppiForMET.puppiCandName    = 'puppiMerged'
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

**Comment by slava77** [comment]
@violatingcp please propagate the fix to 80X, better by updating #15429

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by violatingcp** [comment]
Updated the PR was not merged, so the update is easier in this case.

**Comment by cvuosalo** [comment]
I'm not sure that this PR fixes the problems seen in #15562. I tested workflow 25200.0_ZEE_13 with 500 events against baseline CMSSW_8_1_0_pre11. The test results show that the changes in Puppi jet energy and Et are not reversed by this PR. The other changes seen in #15562 are not reversed either, as seen below.
chetfrac
phoetfrac
met
metlog

**Comment by violatingcp** [comment]
Thats correct. Only the effect of the electrons on the MET will change.

**Comment by cvuosalo** [comment]
In addition, there is a large shuffling of Puppi jet phi values, which seems odd, as well as significant changes in electron and muon Et, as shown below.
phi
elet
muet

**Comment by cvuosalo** [comment]
@monttj: Your review of this PR and #15429 would be helpful.

**Comment by cvuosalo** [comment]
@violatingcp: Please elaborate. Could you please list the problems seen in testing of #15562, of those, which are intended to be addressed by this PR?

**Comment by slava77** [comment]
@violatingcp 
what about the increased tail in puppiMET?
(the plots are for ZEE)
In https://github.com/cms-sw/cmssw/pull/15562 https://cloud.githubusercontent.com/assets/5736159/18070536/1c12bee4-6e14-11e6-888a-09dc10ecf44d.png shows a large increase in MET>100 GeV.
This PR does not change the MET distribution according to https://cloud.githubusercontent.com/assets/5736159/18257541/2e258284-738b-11e6-8443-ef15168f0a98.png

**Comment by violatingcp** [comment]
Why are the events different and what is black in the second plot? The code
should now be exactly the same as what Maria is doing.

**Comment by mariadalfonso** [comment]
The difference between my plot and carl's plot is in the workflow at this point: 
I  made the plots with makePuppiesFromMiniAOD https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/test/corMETFromMiniAOD.py#L118-L128

And Carl is using the makePuppies and https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L249-L258

**Comment by slava77** [comment]
On 9/5/16 4:09 PM, mariadalfonso wrote:
> The difference between my plot and carl's plot is in the workflow at
> this point:
> I made the plots with makePuppiesFromMiniAOD
> https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/test/corMETFromMiniAOD.py#L118-L128
> 
> And Carl is using the makePuppies and
> https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L249-L258

Shouldn't both produce similar results?

If makePuppiesFromMiniAOD is correct, while makePuppies provides
incorrect results,
the makePuppies should be fixed.
If it can not be fixed, then makePuppies should better be disabled: less
confusion and no waste of CPU.
The disadvantage of it though is that validation apparently uses
makePuppies and we lose validation.

**Comment by violatingcp** [comment]
From the PUPPI end the code should do EXACTLY the same thing. I suspect
this might be from the code after Puppi.

Just to make it clear the only substantive difference in the lines is the
setting of UseNoLep as flag which does nothing :
https://github.com/violatingcp/cmssw/blob/bd4dd2d69e962bf71218e142b07631d8b15b876d/CommonTools/PileupAlgos/plugins/PuppiProducer.cc#L204-L205

**Comment by cvuosalo** [comment]
@violatingcp: Just for clarification, in all the plots above, black is the baseline, which is 810pre11 in this case, and red is the baseline plus the PR.

**Comment by violatingcp** [comment]
Ok, so this only includes the merged PR with the one line change. The
previous PR is there.

**Comment by cvuosalo** [comment]
@violatingcp: For my understanding, which of the plots shown in the #15562 thread did you see as indicating a problem, and which of these problems were you trying to fix in this PR?

**Comment by violatingcp** [comment]
So my concern was the MET tail, but it seems as this issue may have nothing
to do with the Puppi code. If that is the case this is a result of
something downstream, which is out of my control. In which case its this
needs to be fixed before a conclusion can be made.

**Comment by slava77** [comment]
@cvuosalo 
Please plot plain slimmedMETsPuppi
(it should be one of the plots made by validate.C in all_mini)

**Comment by cvuosalo** [comment]
@slava77: Are any of these plots what you'd like to see? There were only four plots produced in all_mini: they are phi, pt, px, and py.

met20
met2
all_mini_testpr15669vsorig_zeepuwf25200p0c_patmets_slimmedmetspuppi__reco_obj_pt

**Comment by slava77** [comment]
On 9/5/16 7:14 PM, Carl Vuosalo wrote:
> @slava77 <https://github.com/slava77>: Are any of these plots what you'd
> like to see? There were only four plots produced in all_mini: they are
> phi, pt, px, and py.
> 
> met20
> <https://cloud.githubusercontent.com/assets/5736159/18259717/b44cd922-73ac-11e6-92a4-1cbeae2e7844.png>
> met2
> <https://cloud.githubusercontent.com/assets/5736159/18259716/b4483318-73ac-11e6-9734-73e548427f1b.png>
> all_mini_testpr15669vsorig_zeepuwf25200p0c_patmets_slimmedmetspuppi__reco_obj_pt
> <https://cloud.githubusercontent.com/assets/5736159/18259747/14c107ce-73ad-11e6-9784-3f8e473e18d6.png>

I was looking for the last one, the verbatim plot of slimmetMetsPuppi.
Do I understand it correctly that the last plot is comparing pre11 with
this PR?
How does the comparison look like for the previous PR (where we have
supposedly observed the problem with the MET tail)?

**Comment by cvuosalo** [comment]
@slava77: All the plots I posted above compare baseline 810pre11 with the PR. Here is the slimmetMetsPuppi pT plot from #15562, which compares baseline 810pre10 (black) with PR #15562 (red):

all_mini_testpr15562vsorig_zeepuwf25200p0c_patmets_slimmedmetspuppi__reco_obj_pt

**Comment by slava77** [comment]
@cvuosalo 
thanks
So, the tails are clearly in the slimmedMETsPuppi and not an artefact of DQM plotting selections.
@violatingcp @mariadalfonso please check or confirm that you are OK with this increased tail in MET.

**Comment by mariadalfonso** [comment]
@slava77 @cvuosalo 
Would be possible to have the DQM plot of the #15562 tested on top of the #15565 (merged this morning)?
Phi's changes act on the rawPuppi while here we are comparing the correctedMET that went fixed in the #15565 (as we had various input tags label PF->Puppi )

I think we can close this one 15669 and remove the commit from the corresponding 80X
@violatingcp

**Comment by slava77** [comment]
@mariadalfonso 
please check the comment in 
https://github.com/cms-sw/cmssw/pull/15562/files#r77749405
it is about the line that has already generated significant MET tail in the past and was removed once.
It was reintroduced in #15562 and again we have MET tail generated.
Should it go away?

**Comment by monttj** [comment]
I am sorry I think I am missing the origin of the discussion. I am wondering if there was any presentation concerning this so that I can follow better?

**Comment by mariadalfonso** [comment]
@monttj.
I will summarize today at the 5.30 RECO meeting, in short the situation is the following.

1) the slimmedPuppiMET in release is unusuable as is basically the status as one year ago (wrong tune w/ PU dependency,wrong weight for the photons, various wrong input tags in the MET sequences) 

2) we have now a better tested workflow (used for ICHEP).
I can only test the full workflow miniAOD to miniAOD 
https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/test/corMETFromMiniAOD.py#L118-L128

while slava/carl are testing the AOD -> miniAOD
https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L249-L258

if you can help in checking the difference between the two would be helpful

3) When I take blindly the PR #15562 + miniAOD->miniAOD I  get awesome results.
https://indico.cern.ch/event/557440/contributions/2284506/attachments/1326764/1992006/PuppETvalidation_PRpostICHEP.pdf
Carl instead see bas thing in theMET  tails.

@slava77 @cvuosalo @violatingcp

**Comment by slava77** [comment]
On 9/7/16 3:06 AM, mariadalfonso wrote:
> @monttj <https://github.com/monttj>.
> I will summarize today at the 5.30 RECO meeting, in short the situation
> is the following.
> 
> 1) the slimmedPuppiMET in release is unusuable as is basically the
> status as one year ago (wrong tune w/ PU dependency,wrong weight for the
> photons, various wrong input tags in the MET sequences)

What are the wrong input tags, after
"Puppi MET and MET significance fixes in the MET tool" cms-sw#15600
?

> 
> 2) we have now a better tested workflow (used for ICHEP).
> I can only test the full workflow miniAOD to miniAOD
> https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/test/corMETFromMiniAOD.py#L118-L128
> 
> while slava/carl are testing the AOD -> miniAOD
> https://github.com/violatingcp/cmssw/blob/016ba7029726f71516e71006197f3eabfec68898/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L249-L258
> 
> if you can help in checking the difference between the two would be helpful
> 
> 3) When I take blindly the PR #15562
> <https://github.com/cms-sw/cmssw/pull/15562> + miniAOD->miniAOD I get
> awesome results.
> https://indico.cern.ch/event/557440/contributions/2284506/attachments/1326764/1992006/PuppETvalidation_PRpostICHEP.pdf
> Carl instead see bas thing in theMET tails.
> 
> @slava77 <https://github.com/slava77> @cvuosalo
> <https://github.com/cvuosalo> @violatingcp <https://github.com/violatingcp>
> 

For production purposes we should first make sure that the saved
collections have good performance.
It looks like despite being old, the slimmedMETsPuppi (the only puppi
MET object that goes to miniAOD)
is worse.

**Comment by slava77** [comment]
I'm reusing this open PR thread, although the discussion is really about #15562 

Here is a plot from 
/RelValZEE_13/CMSSW_8_1_0_pre11-PU25ns_81X_mcRun2_asymptotic_Candidate_2016_08_30_11_31_55-v1/GEN-SIM-RECO

using CMSSW_8_1_X_2016-09-06-1100 as a baseline (with all latest from MET, including #15565 from cms-met: Puppi MET and MET significance fixes in the MET tool)
- black: baseline IB
- red: with the questionable line removed https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...slava77:CMSSW_8_1_X_2016-09-06-1100/sign761/puppiPhotonC

all_sign761-pass-1c83cabvsorig-pat-orig_relvalzee_13pu25nsc_patmets_slimmedmetspuppi__pat_obj_pt

**Comment by slava77** [comment]
here are the same kind of plots walking upstream in the slimmed MET production
all_sign761-pass-1c83cabvsorig-pat-orig_relvalzee_13pu25nsc_patmets_patmetspuppi__pat_obj_pt
all_sign761-pass-1c83cabvsorig-pat-orig_relvalzee_13pu25nsc_recopfmets_pfmett1puppi__pat_obj_pt
all_sign761-pass-1c83cabvsorig-pat-orig_relvalzee_13pu25nsc_recopfmets_pfmetpuppi__pat_obj_pt

All these plots bring the performance back, visually, to before #15562 (see Carl's plot in https://github.com/cms-sw/cmssw/pull/15669#issuecomment-245062534 )

the last is the pfMetPuppi made out of puppiForMET using the standard MET producer. puppiForMET is the output of PuppiPhoton. there is no MET code further upstream.

If the purpose of the line of code I'm questioning is to add MET tails in a sample with no intrinsic MET (maybe for better response or something), please clarify.
@mariadalfonso maybe you can check on your side to confirm what @violatingcp said that this line is essential and removing it makes things worse (opposite of what I see)

**Comment by mariadalfonso** [comment]
@slava77 do you know how we can get the references in the PFcandidates

For packed candidate we do like that , but we do not know what to do for "particleFlow"
https://github.com/violatingcp/cmssw/blob/691379a0f452cf21724f83c3978bf13682245d8b/CommonTools/PileupAlgos/plugins/PuppiPhoton.cc#L81

**Comment by slava77** [comment]
On 9/8/16 5:04 AM, mariadalfonso wrote:
> @slava77 <https://github.com/slava77> do you know how we can get the
> references in the PFcandidates
> 
> For packed candidate we do like that , but we do not know what to do for
> "particleFlow"
> https://github.com/violatingcp/cmssw/blob/691379a0f452cf21724f83c3978bf13682245d8b/CommonTools/PileupAlgos/plugins/PuppiPhoton.cc#L81
> 

Do you have to use PF?
To improve reproducibility with running on miniAOD, I'd suggest you use
the same inputs when running on top of AOD.
Are there benefits to use full PF info?

**Comment by mariadalfonso** [comment]
@slava77 @violatingcp 

There are two needs.
1) prepare a function that run with AOD inputs only
2) to run the production AOD --> miniAOD we can just use makePuppiesFromMiniAOD and not use at all the makePuppies. 

This afternoon we can update this PR to work on the direction of 2) as you seems suggesting and post-pone development later

**Comment by slava77** [comment]
On 9/8/16 5:20 AM, mariadalfonso wrote:
> @slava77 <https://github.com/slava77> @violatingcp
> <https://github.com/violatingcp>
> 
> There are two needs.
> 1) prepare a function that run with AOD inputs only

Here I assume you mean "current default PAT" workflow (not the miniAOD
workflow,
which is non-default PAT).
If support for this is needed and is currently enabled in the PAT unit
tests,
then we'd need to stick with possibility to run option 1)

> 2) to run the production AOD --> miniAOD we can just use
> makePuppiesFromMiniAOD and not use at all the makePuppies.
> 
> This afternoon we can update this PR to work on the direction of 2) as
> you seems suggesting and post-pone development later

Please make sure to start with updates of 81X first, to keep it in sync.
Also, before updating the PR, please make sure to run the tests
including "scram b runtests"

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cvuosalo if you follow up with the review, please make sure to check the timing on events with PU (the same ZEE sample should be OK).
Using deltaR can become pretty costly if not done minimally/carefully.

**Comment by mariadalfonso** [comment]
@violatingcp 
With the goal of keep in synch the miniAOD -> miniAOD and AOD->miniAOD, if the DR test is pass then 

1) activate the DR in the makePuppiesFromMiniAOD that is uncomment this lines
https://github.com/violatingcp/cmssw/blob/0b8cf5069f232e9e93e2f40e68452df1d23e7219/PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py#L44-L45

2) switch the full AOD-miniAOD production to makePuppiesFromMiniAOD instead of makePuppies
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L275-L276

@slava77 do you agree ?

**Comment by slava77** [comment]
On 9/8/16 8:20 AM, mariadalfonso wrote:
> @violatingcp <https://github.com/violatingcp>
> With the goal of keep in synch the miniAOD -> miniAOD and AOD->miniAOD,
> if the DR test is pass then
> 
> 1) activate the DR in the makePuppiesFromMiniAOD that is uncomment this
> lines
> https://github.com/violatingcp/cmssw/blob/0b8cf5069f232e9e93e2f40e68452df1d23e7219/PhysicsTools/PatAlgos/python/slimming/puppiForMET_cff.py#L44-L45
> 
> 2) switch the full AOD-miniAOD production to makePuppiesFromMiniAOD
> instead of makePuppies
> https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py#L275-L276
> 
> @slava77 <https://github.com/slava77> do you agree ?

Sounds good.

In interest of doing less harm, in case these two steps have other issues,
please do not update the signed  "Descoped" 80X  PR.
By default we go on with it in 80X.

I guess you can update #15429

**Comment by mariadalfonso** [comment]
@slava77 @cvuosalo let us know how your validation of this extra DR.
if ok then @violatingcp will do the changes discussed in https://github.com/cms-sw/cmssw/pull/15669#issuecomment-245640402 in both #15669 and #15429

#15772 (descoped) will not be modified

**Comment by violatingcp** [comment]
After discussion with @mariadalfonso we have decided #15669 contains all the fixes we would like. The proposal of two additional steps would require a larger re-engineering and consequently we stay with the current #15429 for the 80X release and #15669 for the 81X provided the checks come back with the expected good performance.

**Comment by slava77** [comment]
@cvuosalo we should keep 81X in sync with all new features/bugfixes that were added to 80X.
So, this PR needs to go in pretty soon.

**Comment by cvuosalo** [comment]
+1

For #15669 b92cdaa9576520851f31b6a3d25d48026e478d4e

Puppi photon fixes for #15562.

This PR is related to #15429, #15562, and #15772 .

The code changes match the corresponding changes in #15429, which is already merged. Jenkins tests against baseline CMSSW_8_1_X_2016-09-08-0900 show numerous tiny DQM plot changes, but most are not significant. The slightly larger changes are as expected, with two examples shown below. See #15429, which represents this PR plus #15562 for 80X, for more test results and plots.

From Jenkins workflow 25202.0_TTbar_13:

met
hfmetfrac

**Pull request merged**